### PR TITLE
improv: opencensus mysql

### PR DESCRIPTION
**Pull request by f-hj**
Adding mysql for oc tracing

### CODE REVIEW

File: src/census/plugins/mysql.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+  /**
+   * Patches Mysql operations.
+   */
+  protected applyPatch () {
+    this.logger.debug('Patched Mysql')
+
+    if (this.internalFilesExports.Connection) {
+      this.logger.debug('patching mysql.Connection.createQuery')
+      shimmer.wrap(this.internalFilesExports.Connection, 'createQuery', this.getPatchCreateQuery())
```
Comment (vmarchaud) [inline src/census/plugins/mysql.ts]: need to patch the prototype here too

File: src/census/plugins/mysql.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      this.logger.debug('patching mysql.Pool.getConnection')
+      shimmer.wrap(this.internalFilesExports.Pool.prototype, 'getConnection', this.getPatchGetConnection())
+    }
+
+    return this.moduleExports
+  }
+
+  /** Unpatches all Mysql patched functions. */
+  applyUnpatch (): void {
+    shimmer.unwrap(this.internalFilesExports.Connection, 'createQuery')
```
Comment (vmarchaud) [inline src/census/plugins/mysql.ts]: and unwrap the prototype

File: src/census/plugins/mysql.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    shimmer.unwrap(this.internalFilesExports.Pool.prototype, 'getConnection')
+  }
+
+  private getPatchCreateQuery() {
+    const plugin = this
+    return (original: Function) => {
+      return function (...args: any[]) {
+        var span = plugin.tracer.startChildSpan('mysql-query', plugin.SPAN_MYSQL_QUERY_TYPE)
+        if (span === null) return original.apply(this, arguments)
+        var query = original.apply(this, arguments)
```
Comment (vmarchaud) [inline src/census/plugins/mysql.ts]: replace with `const`

File: src/census/plugins/mysql.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+          query.on('end', function() {
+            span.end()
+          })
+        }
+        return query
+      };
+    }
+  }
+
+  private getPatchGetConnection() {
```
Comment (vmarchaud) [inline src/census/plugins/mysql.ts]: should more something like this :
```
private getPatchGetConnection() {
  const plugin = this
  return (original: Function) => {
    return function (cb: Function) {
      return original.call(this, plugin.tracer.wrap(cb))
    }
  }
}
```

File: src/census/plugins/mysql.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    return (original: Function) => {
+      return function (...args: any[]) {
+        return original.call(this, this.tracer.wrap(arguments))
```
Comment (vmarchaud) [inline src/census/plugins/mysql.ts]: the call to `this.tracer` will be undefined cause `this` refer to the function scope.
arguments will not work too cause `.wrap` accept function and not array

File: src/census/plugins/__tests__/mysql.spec.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
+import { CoreTracer, RootSpan, SpanEventListener, logger } from '@pm2/opencensus-core'
```
Comment (vmarchaud) [inline src/census/plugins/__tests__/mysql.spec.ts]: need test for the connection pool patchs

Comment (vmarchaud) [reply src/census/plugins/__tests__/mysql.spec.ts]: see example: https://github.com/vmarchaud/42_workspace/blob/master/matcha/app/controllers/account.js#L17

File: src/census/plugins/mysql.ts (hunk 1 above)
Comment (f-hj) [reply src/census/plugins/mysql.ts]: Not a prototype
https://github.com/googleapis/cloud-trace-nodejs/blob/master/src/plugins/plugin-mysql.ts#L82
https://github.com/mysqljs/mysql/blob/master/lib/Connection.js#L25

**Pull request merged**